### ISSUE TITLE
Preservation of comments and whitespace

### ISSUE THREAD

**Issue by ned14** [marzer/tomlplusplus#28]
Firstly, this is one of those very few libraries I've used where I don't think I'd have done much different myself. Well, possibly I'd have made it always non-throwing. Well done on that, and thank you for making it open source.

Another thing that I would have done is optional comment and whitespace preservation, such that if you read in a TOML file, and then write it out, comments and whitespace are preserved. This is useful because I'm using your library a bit like how clang-tidy or doxygen do things, where you can dump out a default TOML config, hand edit it, or supply TOML delta files to merge against the default TOML config, and so on. Therefore, dumping out the post-merge TOML file, with all comments also merged, would be ideal.

For this to work right, you'd need to retain comments and whitespace as an attribute of the first node following the comment. Then, if the node gets replaced by a merge, one could retain the existing comment and whitespace, or replace it with the new comment and whitespace etc.

I'd personally make this opt-in, as plenty of people don't want to waste CPU nor memory on comments and whitespace. My thanks in advance to you for your consideration!

**Comment by marzer** [comment]
> Firstly, this is one of those very few libraries I've used where I don't think I'd have done much different myself. 

Damn, that's high praise. Thanks! Glad you like it. It's the first general-purpose C++ library I've released for public consumption so it's been nice to see it well-received.

> Another thing that I would have done is optional comment and whitespace preservation, such that if you read in a TOML file, and then write it out, comments and whitespace are preserved. This is useful because I'm using your library a bit like how clang-tidy or doxygen do things, where you can dump out a default TOML config, hand edit it, or supply TOML delta files to merge against the default TOML config, and so on. Therefore, dumping out the post-merge TOML file, with all comments also merged, would be ideal.

Yeah, this is a piece of functionality I've been asked about a couple of times now, though you're the first to actually provide a compelling use-case. It's tentatively on my mental road-map for `toml++` though might be a decent while before I can make time to implement it.

**Comment by marzer** [comment]
Also, reciprocal praise: your work on profiling the STL headers (e.g. [here](https://www.reddit.com/r/cpp/comments/860hya/stl_header_token_parsing_benchmarks_for_vs2008/)) is something I referred to on a regular basis in optimizing compile times for the library

**Comment by ned14** [comment]
You're very kind. I think if you're a certain kind of C++ dev, stuff like not including headers and, generally speaking, not having the compiler generate opcodes in the first place, leads to a certain shared philosophy. One which I would prefer to see more of on WG21, but that's by the by.

Am I right that your iterators don't implement `operator->`?

**Comment by marzer** [comment]
Quite likely. Most library functionality has been implemented in a very YAGNI way, i.e. I've only added something when I've personally needed it myself. I added the iterators to make ranged-for work and haven't really used them directly via begin()/end() at all.

Which of course means it wasn't an active design decision _not_ to implement the operator, it just hasn't been done. All PR's welcome! :D

**Comment by levicki** [comment]
Let me add my voice towards implementing comment and whitespace (at least `[ \t\r\n]`) preservation.

**Comment by ned14** [comment]
The need to retain comments across modification is becoming more urgent within our use case here, as a wider set of engineers use the tooling. We currently have TOML config files where the original source is commented with documentation describing each option, but once it passes through tomlplusplus all those comments get lost. This is quite unhelpful.

If you even just added the ability for there to be a custom void * per toml table, into which we could store an unordered map of toml identifier to comment, that would be wonderful.

**Comment by marzer** [comment]
@ned14

> If you even just added the ability for there to be a custom void * per toml table, into which we could store an unordered map of toml identifier to comment, that would be wonderful.

Yeah, that'd be doable, though I'd likely make it opt-in to avoid ABI breaks for the masses. What say you to the idea of me exposing it as an "injectable", not unlike ImGui's [IM_VEC2_CLASS_EXTRA](https://github.com/ocornut/imgui/blob/17536f9add3ed3ebdc7fb40d19c6bacb92ea73f8/imconfig.h#L69) and friends?

**Comment by ned14** [comment]
One possible ABI neutral way is an "invisible" node type which doesn't appear in the normal enumerations. It only appears if you explicitly ask for it by name. Said custom node type could carry user defined payload.

I'm not sure if you're already using it, but `node_type::none` could mean just such an invisible node.

**Comment by marzer** [comment]
That value is not presently used in that manner, but I'd rather not go down that path.

To be clear, I'm not at all concerned about ABI breaks for those opting in to this feature; it would be "buyer beware" advanced use case. I'm only worried about avoiding them for those not opting in.

**Comment by ned14** [comment]
I assume `node_type::custom` is also not your preference?

**Comment by marzer** [comment]
If I were to implement this fully I'd likely treat the comments as metadata associated with the tables/arrays/values they relate to, similar to how the source information works now, rather than add them as nodes.

But, er, unless I misunderstood what you were asking for as an interim solution, you suggested adding a public `void*` data member to `toml::table`, right? In which case I'm saying that I'd be happy to go a little further and allow advanced users to inject whatever they like using a config option. The only caveat for either route would be that I'd gate it behind an opt-in configurable so as to not break ABI for everyone else.

**Comment by ned14** [comment]
I know that toml11 handles comments by attaching them to nodes, but that falls down if there are blocks of comment without any obvious node to attach them to. What most parsers do in this situation is to have comment nodes which act like whitespace i.e. ignored by default.

But separately to all that, what would be ideal for me right now is to be able to store custom data with a secret named key. Then I can do a poor man's regex parse of the incoming TOML, and say for a key named `foo` I might store an invisible custom node called `#comment_foo` with the comment text. When reconstituting the TOML for output as text, I can then inject the per-key comments from the original.

I know this probably sounds like I'm just going ahead and implementing comments, but because our TOML config files are super simple, I can hack a nasty solution which will work for our simple TOML. It wouldn't likely scale out well.

Failing that, the custom void * per toml table can also work. Just be aware that we may see mixed `toml::table*` in our code base because code A might create the custom table pointers, but code B might create default ones. So we'd need a way to disambiguate between them.

I originally thought to simply subclass `toml::table` with a custom type, then use RTTI to disambiguate, but you made `toml::table` final :(

**Comment by levicki** [comment]
Guys, let's not rush this and make some convoluted implementation which isn't useful for everyone interested.

For example, I would really like comments preserved verbatim with their relative positions to nodes/tables:
```
# where this comment belongs?
# and this one?

# how about this one?

title = "Example" # or this one?

# and now this...
[bcde] # or this?
# or even this?
key1 = value1 # welp...

# maybe it's gremlins?

# no way
[abcd] # oh crap...
# what the...
key1 = value1 # not again...

# or the end of the file?
# shurely shome mishtake?
```
IMO, the only way to preserve those is to keep the relationship to tables and key/value pairs and shuffle them together when sorting the output on write.

One way to accomplish that is to have one fictional node called BOF (as in Beginning Of File) and all comments that come after file start but before first table or array of tables should be kept relative to BOF (i.e. as its children nodes).

As for the rest of comments, inline comments should not be too problematic. The rest of them should be kept either relative to the table or array of table or to the key/value pairs. Given the TOML rule that every kvp after `[table]` or `[[table]]` belongs to that node until next node or EOF is observed it shouldn't be too hard to apply the same rule to the comments as well.

**Comment by marzer** [comment]
@ned14 

> I know that toml11 handles comments by attaching them to nodes, but that falls down if there are blocks of comment without any obvious node to attach them to. What most parsers do in this situation is to have comment nodes which act like whitespace i.e. ignored by default.

Yeah, it's not a perfect solution, but it's along the lines of what I'd do, too. Another problematic case is comments at the end of a TOML document, though in that case I'd probably just associate them with the most recently seen node. The rest can be glommed together and associated with the preceding element al a doxygen markup and similar:

```toml
# this comment would be assoicated with "key"
key = "value"

# this comment would be associated with "table"
[table]

	# this comment would be associated with "table.subtable"
	
	# so would this one
    [table.subtable]
	
   # ...and so would this one. This one doesn't really make sense, but what does here, really?
```

Obviously it's not perfect, but it would do the job for the majority of cases I'd think. It also has the practical upshot of maintaining semantic relationships between comments and toml data, whereas if they were two discrete node types, that relationship would be lost, as @levicki noted above.

The alternative is a full DOM-style rendition of the TOML with whitespace and comments all treated as separate nodes in the tree, but that seems wholly overwrought for what is fundamentally just a set of key-value pairs, and is just not a direction I want to go in with this library.

Regardless of how it's implemented, I won't be able to attack it for a while still, but I'm happy to patch in a simpler hook for you to add a tag pointer or something in the interim.

----

@levicki 

> One way to accomplish that is to have one fictional node called BOF (as in Beginning Of File) and all comments that come after file start but before first table or array of tables should be kept relative to BOF (i.e. as its children nodes).

Yeah that's one way of doing it. See my example above for another. There's lots of wiggle room here, and no single approach is perfect.

> Please don't allow them to inject arbitrary stuff because that might enable breaking the TOML spec and allow producing non-compliant TOML files.

I meant injecting stuff into the `toml++` node classes, like a `void*` tag member or similar. Nothing to do with the emitted TOML when re-serializing etc.

> I like your idea of comments being metadata, it would be nice if there was a proper interface to add/modify/remove comments.

If I went down this route I'd certainly make them mutable. It wouldn't be much of a stretch.

> I think they should be preserved by default, and there should be a flag on writing to strip them out for those who don't need them.

In the absence of concrete data I can only speculate, but I suspect the considerable majority of people who go looking for a TOML library do so because they want their application to read in a set of key-value pairs from a config, with serializing coming second and needing to mess with comments being a distant third. The library will always be structred with this in mind, so I won't ever make stuff like that the default.

**Comment by levicki** [comment]
Mark,

In my opinion, your example of where comments belong does not fit well with TOML spec.

First, any KVPs between start of file and the first table/array of tables belongs to global scope and so should any comments there -- in other words, if a KVP preceding a table does not belong to it, then the comment  preceding a table shouldn't either.

Same goes for this:
~~~
[table1]
# comment1
Key=value
# comment2
[table2]
...
~~~
Like it or not, comment2 belongs to table1, not table2. Proper way to comment table2 is to put comment inline or right below it.

Finally, I am not sure if TOML serialization requires sorting the tables and KVPs aplhabetically, but if that is not necessary then not changing order on serialization would go a long way towards allowing you to preserve the relationship between the comments and data.

I would disabke serialization sorting and make each line a node. Inline comments would be metadata on respective node type, comment lines would be empty nodes with metadata, and finally empy lines would be empy nodes with empty metadata. That way you could preserve structure, and sorting would not mess with diffs of before/after TOML files as they currently do.

**Comment by marzer** [comment]
> In my opinion, your example of where comments belong does not fit well with TOML spec.
> Proper way to comment table2 is to put comment inline or right below it.

Well I disagree with that outright. You might like to think of comments as structures in the TOML tree, but that's not how comments are typically used. Typically a comment precedes the thing it relates to. Sure there's exceptions to this, but I bet I'm in the majority on this. The TOML spec is irrelevant here.

As for the rest of it, you've basically just described a DOM. I'm not going in that direction.

**Comment by levicki** [comment]
Yes, humans usually put comments before.

However, given this:
~~~
# preceding comment
[business] # inline comment
Text = "he means it"
~~~

How do you attach both comments as metadata to a single table? How do you specify which one you want to change or delete? 

What happens if you have this:
~~~
# preceding comment

[business] # inline comment
Text = "he means it"
~~~

Or this:
~~~
# preceding comment
[[business]] # inline comment
Text = "he means it"

[[business]] # inline comment

[[business]] # inline comment
Text = "as usual"
~~~
I am curious as to what your solution to that would be.

**Comment by marzer** [comment]
I just want to take a moment to re-frame the discussion, because the potential for scope-creep is high.

Ultimately for me to implement this in an order-preserving, whitespace-preserving, comment-preserving, DOM-like fashion, would require effectively re-writing the whole library, adding a bunch of bloat, and making it much, much slower. Not to mention the time and energy!

So, suffice it to say that I will _never_ be doing that. That's not what I want for this library. That's a different library altogether. Anything I implement in aid of closing this feature request will be a "best-fit" compromise.

I can tell you outright. I'm not ever going to bother preserving whitespace. Comments, sure, to some extent, but not whitespace.

I _might_ consider preserving order at some point, but that would have pretty bloaty knock-on effects.

So, with that in mind, let's focus on just documentation comments to begin with, since that seems the most generally useful.

**Comment by ned14** [comment]
For me personally:

```
[table1]
# comment1
Key=value
# comment2
[table2]
...
```

In my opinion, comment2 is a comment about table2, and does not belong to table1. It is absolutely the case that one would want to comment on a whole table, as well as on individual keys in a table.

Re @marzer's wish to not go full DOM, there are actually ways of implementing primarily a key-value store which is very fast and noncopying, and which preserves comments and whitespace, and avoids dynamic memory allocation and is never throwing. https://github.com/chadaustin/sajson would be the classic example, it mutates the original text save on memory allocations and memory copying. But he's right it's a complete rewrite, total rearchitecture.

Ultimately I can live with per node comments if I also have the ability to keep a void* per table.

**Comment by marzer** [comment]
@ned14 Yeah it seems your model for relating comments to content is the same as mine (and most coders more generally, I suspect). Should be easy to come up with something that works well for those cases.

It seems there's some urgency here so I'll try to come around to this in a reasonable timeframe, but this year has taken a toll on my productivity in a lot of unexpected ways...

**Comment by h-vetinari** [comment]
Hey all, thanks for this library (which I'm working with from the [python side](https://github.com/bobfang1992/pytomlpp)) and this discussion! :)

I agree with many of the points that were said already, but I want to bring in #62 more specifically again - I think it's a fool's game to try to guess which line a comment belongs to, as people have different conventions where something makes most sense, sometimes even in the same file.

> In fact, the TOML spec explicitly states that order doesn't matter, so the library is written with this in mind (i.e. things are serialised in implementation-defined order, because order has no meaning).

The larger issue is that TOML is specifically designed to be both human and machine-readable (and the spec cares mostly about the machine side of things, while human interaction patterns are... more complicate). I'd argue that while the spec (and `std::map`) doesn't care about the order, humans do, _a lot_. Therefore, to do automated manipulation of a human-maintained toml-file, we should strive for idempotency of a read-write-loop (at least, once the TOML is in a given canonical format).

I know that this would make the implementation more complicated, possibly less performant, etc. But for a production-grade TOML library (which might take a while to evolve across the ecosystem), it's IMO a crucial piece. The good thing is that this library already has an abstraction layer over `std::map`, so it should be possible to optionally make this ordered as well - some ideas collected [here](https://stackoverflow.com/questions/1098175/a-stdmap-that-keep-track-of-the-order-of-insertion), for example.

Assuming can keep the order of nodes (or at least have an option for that), then it becomes mostly arbitrary if we attach comments to the node before or after (just needs a special node at the end for either the first/last comment).

**Comment by levicki** [comment]
@marzer I think I owe you an explanation.

> I suspect the considerable majority of people who go looking for a TOML library do so because they want their application to read in a set of key-value pairs from a config, with serializing coming second and needing to mess with comments being a distant third.

I came upon your library and started using it in my projects exactly because it offered serialization unlike the other libraries that were good enough for just reading key-value pairs.

It is one thing to use library to read key-value pairs and I agree that most people care about that (but then any other TOML parser will do, not just yours).

However, imagine situation where you have a bunch of services without user interface all using a TOML config file on say a bunch of ATMs scattered in remote locations. Imagine that config options are documented so that people who have to change them when doing field maintenance can have a clue what they are doing. Imagine also that sometimes you need to make an automated change on many config files remotely.

You cannot make changes in text editor and just replace all existing config files because each config file has machine-specific configuration. So what are you options?

1. Downloading, editing, and re-uploading 100 config files to preserve comments? Too time consuming and error-prone.
2. Writing a command line tool that will make the change by properly serializing each config file on each machine? Impossible, unless you want to destroy comments and change key sort order.

Same problem if you have a service that reads the config file and a GUI application that writes and modifies said file. If you need to change anything remotely, you need that config file to have comments. However, the next time someone runs that GUI application when doing field maintenance your comments will be gone.

I am not aware of any TOML parser that can support those two in my opinion totally valid use scenarios for TOML config files. Yours was the closest until...

> The library will always be structred with this in mind, so I won't ever make stuff like that the default.

So with that in mind...

> it's because this part was actually a dialogue with @levicki, who appears to have since deleted his comments. I don't know why. 

Well I guess you know why now -- you have made your mind and staying in that discussion seemed pointless.

**Comment by ned14** [comment]
> For me personally:
> 
> ```
> [table1]
> # comment1
> Key=value
> # comment2
> [table2]
> ...
> ```
> 
> In my opinion, comment2 is a comment about table2, and does not belong to table1. It is absolutely the case that one would want to comment on a whole table, as well as on individual keys in a table.

So, the above changed recently in our work codebase because the TOML file got big, and the ops team wanted more use of horizontal screen estate, so now we're doing:

```
[table1] # table explanatory comment
key1=value # key1 explanatory comment
           # second line of key1 explanatory comment, note the whitespace alignment
```

Because our TOML is nice and straightforward, I have this awful munge piece of code that serialises the TOML, does some regex matching and patches in the explanatory comments for the keys before dumping to disc. It is slow, buggy, and in every way unsatisfactory.

In any case, comment-next-line-node and node-same-line-comment both seem valid orientations.

**Comment by marzer** [comment]
@h-vetinari

> I'd argue that while the spec (and std::map) doesn't care about the order, humans do, a lot. Therefore, to do automated manipulation of a human-maintained toml-file, we should strive for idempotency of a read-write-loop (at least, once the TOML is in a given canonical format).

Eh. I disagree. The points I've outlined earlier in this thread cover why pretty comprehensively. _Some_ humans care about that a lot, but many others do not. That being said, I would still like to put something in place to allow this for the sake those who would like it (e.g. @ned14), but any efforts to do so are going to necessitate a pretty comprehensive rewrite and I just don't have the time to do that currently.

Plus, if I'm being honest, I'm pretty creatively burned out on this project at this point. Motivation comes and goes, but all I've been able to muster over the last 6 months or so is bugfixes and small refactors. I might get a bolt of energy for a big batch of overhaul/feature work at some point, but I don't want to promise anything.

----

@levicki 

> Well I guess you know why now -- you have made your mind and staying in that discussion seemed pointless.

That makes sense, I guess, but deleting all your correspondence (making the other parties look insane in the process) is not the same as simply bowing out of a discussion O_o

**Comment by h-vetinari** [comment]
@marzer answered on the issue from the tracker of the python-wrapper that I linked - I'm choosing to answer here:

> > @h-vetinari: [...] fixing it on the tomlplusplus side [...]
> 
> @marzer: Fixing it implies that it's broken. It's not. It disregards order on purpose, as I justify quite comprehensively in the thread you linked. If you want this 'fixed', fix it in python, as toml++ will stay the way it is for the forseeable future.

That's... very unfortunate. Do I understand you correctly that you would not even consider/accept a PR that makes this optional on the `toml::table`?

> Eh. I disagree. The points I've outlined earlier in this thread cover why pretty comprehensively. Some humans care about that a lot, but many others do not.

I read the thread, but I'm not so sure, since it effectively precludes all use cases that involve modifying a toml-file that's _also_ editable by a human. And that's not a fringe use case at all, because readability and meta-information like comments are the key differentiator of TOML vs. JSON. And while most use-cases start out as "I need to read this data", the choice of format already signifies that human interaction is an important requirement, and automated reprocessing (if only for validation of the human edits) is something that almost always comes up soon afterwards.

In fact, I'm not aware of a toml library that is able to do this for now, so it becomes a self-fulfilling prophecy that writing comes as an afterthought in use-cases (or rather, everyone has to swallow that they cannot do that automated reprocessing of human-edited files).

> Plus, if I'm being honest, I'm pretty creatively burned out on this project at this point. Motivation comes and goes, but all I've been able to muster over the last 6 months or so is bugfixes and small refactors. I might get a bolt of energy for a big batch of overhaul/feature work at some point, but I don't want to promise anything.

Sorry to hear - and very understandable; OSS maintainership is difficult at the best of times, much less the current ones.

**Comment by marzer** [comment]
@h-vetinari I'd happily accept a pull request if you want to make it configurable with some drop-in map replacement, but do be aware that it needs to support heterogeneous lookup; last time I tried to find some insertion-order-tracking map with that capability I came up empty handed.

Failing that, though, that machinery would need to be added to `toml::table` itself, which is something I'd like to avoid.

**Comment by h-vetinari** [comment]
> I'd happily accept a pull request if you want to make it configurable with some drop-in map replacement, but do be aware that it needs to support heterogeneous lookup

OK - not sure if/when I'll find time myself, but glad to hear that the door is not fully closed

> last time I tried to find some insertion-order-tracking map with that capability I came up empty handed.

Mostly leaving that as comment for myself so I don't forget - this [SO question](https://stackoverflow.com/questions/64673031/implementing-heterogeneous-lookup-for-spans-in-c) has some thoughts about this based on abseil's `flat_hash_map`. Some further [reading](https://abseil.io/docs/cpp/guides/container) also shows that their [`btree_map`](https://github.com/abseil/abseil-cpp/blob/20210324.1/absl/container/btree_map.h#L58-L69) might be an interesting source of ideas.

**Comment by marzer** [comment]
> glad to hear that the door is not fully closed

Well truthfully I don't consider anything to be fully closed in an absolute sense; I'd _love_ toml++ to be the fastest, most fully-featured TOML library on the planet; the only blocker is this meagre human form and it's time-bound constraints, heh. I have to draw lines and avoid making commitments on things, and that usually ends up favouring the status quo (as you pointed out above).

The problem is lessened significantly when it's not me doing the work :D

**Comment by pwm1234-sri** [comment]
Let me first say that toml++ looks like a great library - thank you for making it available!

But my two cents on this issue is that reasonable preservation of comments is a requirement for me for working with toml. I have the exact same use case as @ned14 described when he opened this issue. I am using toml as a config format for another library, which outputs toml with comments, and I need to tweak its values or add new values without removing the original comments. I look forward to toml++ supporting preservation of comments, but until it does, I am using toml11 because it has support for preservation of comments.

**Comment by marzer** [comment]
@ned14

> I originally thought to simply subclass toml::table with a custom type, then use RTTI to disambiguate, but you made toml::table final :(

I've started some refactoring and cleanup work on the [v3 branch](https://github.com/marzer/tomlplusplus/tree/v3), and in doing so have removed `final` from the value, array and table classes (instead moving the `final` keyword to the overrides so I can still help the devirtualizer). 

@pwm1234-sri 

>  look forward to toml++ supporting preservation of comments, but until it does, I am using toml11 because it has support for preservation of comments.

Good to hear that toml11 has the feature you need. It's a good library and the author is pretty active in maintaining it.

**Comment by julianxhokaxhiu** [comment]
Hello,

First of all thank you very much for this awesome library. I've been using for quite some time now on top of my own project and I'm really happy about it. As many others here I also now ended up clashing on this feature request and at least I can explain my own use case, hoping it helps to drive this implementation forward.

On my own case I have [this toml file](https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.lighting.toml) that contains a simple explanation on how the file will be used, but I'll extend this to have something [like this other toml file](https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.music.toml#L3-L45). If I could preserve this part while parsing/saving it back it would be fantastic.

Is there any way I could approach this even with the current status of the library? I'll be happy to implement it on my own side if you can please let me know how to achieve such a thing.

Thank you in advance until then.

**Comment by marzer** [comment]
@julianxhokaxhiu 

> as apparently this library doesn't support in-memory editing of the parsed toml

Huh? It absolutely does. There are many examples throughout the documentation.

As for the rest of your query, no there's no built-in way to do that. My solution to that would be to keep the comment-based instructions in a separate file and concatenate the formatted toml together with the template before writing them to a file.

**Comment by julianxhokaxhiu** [comment]
> Huh? It absolutely does. There are many examples throughout the documentation.

I've seen only [this one](https://marzer.github.io/tomlplusplus/#mainpage-example-manipulations) but it doesn't help my own use-case. Apparently any tentative i do to manipulate data once I parse a file, it doesn't work. The compiler refuses to continue as it doesn't recognize my left side assignment, if this makes any sense.

> As for the rest of your query, no there's no built-in way to do that. My solution to that would be to keep the comment-based instructions in a separate file and concatenate the formatted toml together with the template before writing them to a file.

The problem I see with this approach is that it will preserve only a portion of it, but not all. For eg. [on this case](https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.toml) what would be the template? I have comments at the beginning and for each flag. I'm not sure how [Tomlyn](https://github.com/xoofx/Tomlyn) manages to handle it, but it does it quite right, as I'm using this library on [this project](https://github.com/tsunamods-codes/7th-Heaven) which manipulates my `FFNx.toml` file quite fine, preserving it visually as you see it, yet updating fields to their relative values assigned programmatically.

If tomlplusplus could achieve the same result, it would make it the final library for C++!

**Comment by marzer** [comment]
> The compiler refuses to continue as it doesn't recognize my left side assignment

Because that's not how you inject additional values using this library. `operator[]` gets you a `toml::node_view`, which is essentially just a pointer with some extra syntax sugar (so you can chain them together, [like this](https://marzer.github.io/tomlplusplus/classtoml_1_1node__view.html)).

If you want to add new elements to arrays or new KVPs to tables, you need to first cast the node to the actual underlying container type and manipulate it that way - see the [[Working with TOML data](https://marzer.github.io/tomlplusplus/#mainpage-example-manipulations)](https://marzer.github.io/tomlplusplus/#mainpage-example-manipulations) example from the main page.

> If tomlplusplus could achieve the same result

I'm not going to be implementing this any time soon, unfortunately.

**Comment by julianxhokaxhiu** [comment]
I see thanks nevertheless. If I may I found another issue as I'm using the correct method you pointed to work with the TOML data structure. As you can see, the first `insert_or_assign` block is totally overridden by the second one. How can I "append" keys to the same table structure? I need to do them at two different execution time so I can't "join" the two statements ( even though in the example they might be joined ).

https://godbolt.org/z/P6c8n99nc

Thank you in advance!

**Comment by marzer** [comment]
I don't understand what you're trying to accomplish there; `table::insert_or_assign()` has the same semantics as `std::map::insert_or_assign()`, so calling it twice with the same key would naturally replace the value inserted via the first call with the value assigned by the second.

If you want to append to the table you created in the first call, get an iterator to it (returned yb the first call to `insert_or_assign`), cast that to the target type, then call whatever table modification operation you need on _that_ child table (e.g. `insert()`).

As a general rule I've followed all of the idioms of the `std` containers where possible, i.e.:
- `std::vector` -> `toml::array`
- `std::map` -> `toml::table`
(the only exception being `toml::table::operator[]` getting a node_view, but that's a price I pay happily because `std::map::operator[]` is a bug factory)

**Comment by julianxhokaxhiu** [comment]
Ok makes sense, thank you :)

**Comment by marzer** [comment]
Example:

```cpp
  config.insert_or_assign("hello", toml::table{
    { "key1", true }
   });

	toml::table* child = config["hello"].as_table();

	child->insert_or_assign("key2", false);
```

Since the casting stuff is bit tedious it's usually easier to build your configs bottom-up where possible, e.g.:

```cpp
toml::table hello{{ "key1", true }};

// ...

hello.insert_or_assign("key2", false);

// ...

config.insert_or_assign("hello", std::move(hello));
```

**Comment by julianxhokaxhiu** [comment]
Works like a charm, just tested it now, thanks a lot! Looking forward to comment preservation hoping you'll find some time to implement it. Until then, thanks a lot of this amazing library!

**Comment by mchaptel** [comment]
Hello, and thanks for this library which has let me get up and running very quickly!

I read a little the previous discussion but couldn't figure out where it landed. Are certain types of comments supported now? Is it considered a feature creep area which will likely never see the day?

I'm asking because, like the original poster, I'm looking to parse, edit and dump a config file and I would like to guide users a little on what setting does what using comments. So any comments support would be amazing, and I'm ok with using a given convention laid out by the lib on what gets kept and what doesn't. 

I have first tried toml11 but I'm facing an issue as that lib breaks the toml spec on dot notation access. This one however has been very straightforward and I feel is more actively supported, so I would like to keep using it.

Is there any update on this issue?

**Comment by marzer** [comment]
> Are certain types of comments supported now? 

Not currently.

> Is it considered a feature creep area which will likely never see the day?

Very likely, yes. I don't really have the time to work on this library much any more, outside of bug fixes. Implementing this necessitates a very large re-write.

**Comment by mchaptel** [comment]
thanks for the reply! 

A shame but I understand. Maybe if I'm feeling it I'll try to make a PR!

**Comment by mchaptel** [comment]
@marzer Sorry to bother, this isn't exactly related to this issue but since I was there I thought I might as well ask.

I have not been able to figure out how to use insert_or_assign with a path (containing dot notations). Nor have I been able to find how I would go about assigning a value to a node retrieved with at_path(). Is there something I'm missing? I'm only trying to assign simple string and float/integer values.

I have looked around the examples and couldn't find one where the key wasn't a root key, or retrieved with an iterator. I'm very confused.

I would greatly appreciate your help!

**Comment by marzer** [comment]
@mchaptel there's no way of doing that, nor will there be. See [this discussion](https://github.com/marzer/tomlplusplus/issues/160) for an explanation.

If you want to modify the document, you need to work with the hierarchy directly.

**Comment by mchaptel** [comment]
thanks for the fast reply! I have found the way you showed above to get the parent as_table then assign afterwards, which might work for my purpose.

**Comment by hayguen** [comment]
i simply workaround by writing following c++ - if there's no existing config file:
```
auto tbl = toml::table{ {
      { "# comment1", "info: keys starting with a '#' are interpreted as comments and ignored" },
      { "# comment2", "this file is automatically created as example/template, showing possible keys" },
      { "# enable", "thus, set enable to 'true' to activate; keep 'false' to deactivate" },
     ...
```

the written file then contains:
```
"# comment1" = "info: keys starting with a '#' are interpreted as comments and ignored"
"# comment2" = 'this file is automatically created as example/template, showing possible keys'
"# enable" = "thus, set enable to 'true' to activate; keep 'false' to deactivate"
```

sure, that isn't perfect .. but is good enough for me.
is there a simpler workaround?

**Comment by marzer** [comment]
@hayguen 

> is there a simpler workaround?

Nope, not currently. Since toml++ does not regard comments and whitespace with any significance, there's nothing to preserve, so I guess your solution is as a good as any, heh.

**Comment by alexchandel** [comment]
Comment/style preservation would be a nice feature to have, as TOML becoming a dominant configuration file format, and programs frequently need to do writebacks.

Python's [tomlkit](https://github.com/python-poetry/tomlkit) (supporting TOML v1.1.0 now) may serve as a good example for how to do this.
